### PR TITLE
dep: update tough version to v0.16.0

### PR DESCRIPTION
**Pull request by Xynnn007**
Manual for #317.

New version of `tough` brings async impls and also a bug which I am fixing in upstream.

Please DO NOT merge this until the upstream `tough` merges https://github.com/awslabs/tough/pull/737

### CODE REVIEW

File: src/tuf/mod.rs [hunk 1]
```diff
@@ -287,14 +295,22 @@ fn fetch_target_or_reuse_local_cache(
...
         )),
-        Some(reader) => {
-            data = read_to_end(reader)?;
+        Some(mut stream) => {
+            let mut data = vec![];
+            while let Some(d) = stream.next().await {
+                let mut d = Into::<Vec<u8>>::into(d.map_err(Box::new)?);
+                data.append(&mut d);
+            }
+            println!("{data:?}");
```
Comment (flavio) [inline src/tuf/mod.rs:313]: nitpick: a leftover, to be removed or turned into a `debug!` statement

File: src/tuf/mod.rs [hunk 2]
```diff
@@ -84,31 +90,32 @@ impl Repository for ManualRepository<'_> {
 pub struct SigstoreRepository {
     repository: tough::Repository,
     checkout_dir: Option<PathBuf>,
-    trusted_root: OnceCell<TrustedRoot>,
+    trusted_root: OnceLock<TrustedRoot>,
```
Comment (flavio) [inline src/tuf/mod.rs:93]: Maybe we should use a [`tokio::sync::OnceCell`](https://docs.rs/tokio/latest/tokio/sync/struct.OnceCell.html) instead?

File: src/tuf/mod.rs
Comment (flavio) [inline src/tuf/mod.rs:1]: I think we have to update the inline docs in these points:

* https://github.com/sigstore/sigstore-rs/blob/315d5f676cbaf83d8c1225078524cfc02f89b837/src/tuf/mod.rs#L186-L187
* https://github.com/sigstore/sigstore-rs/blob/315d5f676cbaf83d8c1225078524cfc02f89b837/src/tuf/mod.rs#L210-L211

Now the blocking operations should have be gone. Am I right?

**Review by flavio — CHANGES_REQUESTED** [review]
@Xynnn007 thanks for this PR and for the upstream fix.

I've left some comments in the meantime

File: src/tuf/mod.rs (hunk 1 above)
Comment (Xynnn007) [reply src/tuf/mod.rs:313]: Oh. Thanks! Let me delete that

File: src/tuf/mod.rs (hunk 2 above)
Comment (Xynnn007) [reply src/tuf/mod.rs:93]: Nice suggestion. Let me try this

File: src/tuf/mod.rs
Comment (Xynnn007) [reply src/tuf/mod.rs:1]: Yes. I have double checked comments upon other modified functions. Thanks for pointing out.

**Comment by Xynnn007** [comment]
@flavio Thanks for the reviewing. I have fixed them. PTAL

**Review by flavio — COMMENTED** [review]
LGTM, waiting for the fix done to tough to be merged upstream

**Comment by Xynnn007** [comment]
Merged in upstream. Still need to wait for a release

**Comment by lukehinds** [comment]
Still no new version as yet, would you care to poke the folks upstream again @Xynnn007 ?

**Comment by Xynnn007** [comment]
> Still no new version as yet, would you care to poke the folks upstream again @Xynnn007 ?

Yes. Pinged in https://cloud-native.slack.com/archives/C8NMD3QJ3/p1708754342493549

**Comment by astoycos** [comment]
Hiya @Xynnn007 I took a stab at rebasing this against main (https://github.com/sigstore/sigstore-rs/compare/main...astoycos:sigstore-rs:main) so that I could bring in these blocking changes -> https://github.com/awslabs/tough/pull/755

But I'm a bit stuck on 
```rust
error: future cannot be sent between threads safely
   --> src/trust/sigstore/mod.rs:168:65
    |
168 |       async fn fulcio_certs(&self) -> Result<Vec<CertificateDer>> {
    |  _________________________________________________________________^
169 | |         let root = self.trusted_root().await?;
170 | |
171 | |         // Allow expired certificates: they may have been active when the
...   |
182 | |         }
183 | |     }
    | |_____^ future created by async block is not `Send`
    |
    = help: within `SigstoreTrustRoot`, the trait `Sync` is not implemented for `std::cell::OnceCell<TrustedRoot>`, which is required by `{async block@src/trust/sigstore/mod.rs:168:65: 183:6}: std::marker::Send`
    = note: if you want to do aliasing and mutation between multiple threads, use `std::sync::OnceLock` instead
note: captured value is not `Send` because `&` references cannot be sent unless their referent is `Sync`
   --> src/trust/sigstore/mod.rs:168:28
    |
168 |     async fn fulcio_certs(&self) -> Result<Vec<CertificateDer>> {
    |                            ^^^^ has type `&SigstoreTrustRoot` which is not `Send`, because `SigstoreTrustRoot` is not `Sync`
    = note: required for the cast from `Pin<Box<{async block@src/trust/sigstore/mod.rs:168:65: 183:6}>>` to `Pin<Box<dyn Future<Output = Result<Vec<CertificateDer<'life0>>, SigstoreError>> + Send>>`
    = note: the full name for the type has been written to '/home/astoycos/go/src/github.com/sigstore/sigstore-rs/target/debug/deps/sigstore-00d994436acc6b0d.long-type-9205612641815476395.txt'
    = note: consider using `--verbose` to print the full type name to the console

error: future cannot be sent between threads safely
   --> src/trust/sigstore/mod.rs:189:54
    |
189 |       async fn rekor_keys(&self) -> Result<Vec<&[u8]>> {
    |  ______________________________________________________^
190 | |         let root = self.trusted_root().await?;
191 | |         let keys: Vec<_> = Self::tlog_keys(&root.tlogs).collect();
192 | |
...   |
199 | |         }
200 | |     }
    | |_____^ future created by async block is not `Send`
    |
    = help: within `SigstoreTrustRoot`, the trait `Sync` is not implemented for `std::cell::OnceCell<TrustedRoot>`, which is required by `{async block@src/trust/sigstore/mod.rs:189:54: 200:6}: std::marker::Send`
```

Can you help out with this?

**Comment by astoycos** [comment]
Well I was able to get things building there ^^ but now 

```rust 
error[E0277]: `dyn sigstore::tuf::Repository` cannot be shared between threads safely
```

Obv now those types referenced above are not `Send` I need to figure out what changed in `tough` to cause this change :/

**Comment by astoycos** [comment]
Rebased and updated this in https://github.com/sigstore/sigstore-rs/pull/340 hope that's alright @Xynnn007!

**Comment by Xynnn007** [comment]
Hi @astoycos I was once planning to update this PR after the upstream `tough` made a new release as we will not merge any PR involving a dep specified by `rev` rather than a `version`.

If you'd like to take this, let me close the PR and assign #317 to you.

**Comment by astoycos** [comment]
Sure @Xynnn007 I'm happy to carry this through if that's alright with you :) 

I'd like to start contributing to sigstore!

**Comment by Xynnn007** [comment]
@astoycos Welcome! Seems that I cannot find you in the asignees. But that does not matter, just go.

**Comment by astoycos** [comment]
Sounds good If you could give https://github.com/sigstore/sigstore-rs/pull/340 a look to make sure I stated this correctly I'd appreciate it :)

**Pull request closed without merging**